### DISCUSSION TITLE
brew test-bot fails in audit

### DISCUSSION THREAD

**Discussion by choppsv1** [Homebrew/discussions#1585] [Tap maintenance and Homebrew development]
I have a tap that builds a modified emacs-mac and it's failing like so in the automated build scripts on github
```
(default) [18:02:17 ja:~]$ brew audit emacs-mac
choppsv1/emacsmacport/emacs-mac:
  * Emacs Lisp files were linked directly to "/opt/homebrew/share/emacs/site-lisp".
    This may cause conflicts with other packages.
    They should instead be installed into:
      /opt/homebrew/opt/emacs-mac/share/emacs/site-lisp/emacs-mac
    The offending files are:
      /opt/homebrew/opt/emacs-mac/share/emacs/site-lisp/subdirs.el
Error: 1 problem in 1 formula detected
```
I checked and the regular emacs.rb formula in homebrew/core installs this same link. How do I get around this audit failure so that my bottles get published?

**Comment by gromgit** [comment]
The formula named `emacs` gets to skip this check, since it's the base formula and so can't conflict with itself: https://github.com/Homebrew/brew/blob/master/Library/Homebrew/formula_cellar_checks.rb#L152-L170

**Reply by choppsv1** [reply]
I see, so I would need to name my formula emacs.rb instead of emacs-mac.rb then.

**Reply by gromgit** [reply]
Or maintain `emacs-mac` and ignore the audit, or do something else. The most appropriate course of action depends on what you're trying to do with `emacs-mac`, which you haven't mentioned.

**Reply by choppsv1** [reply]
Sorry, https://github.com/choppsv1/homebrew-emacsmacport has the description.. I'm renaming the formula now.

**Reply by gromgit** [reply]
Note that if you rename your formula to `emacs`, you're forcing a namespace clash with Homebrew core that may have unpredictable consequences. Since you're hosting a third-party tap, the Homebrew maintainers can't really do much to stop you, but don't be surprised if a future `brew upgrade` replaces your enhanced `emacs` with the stock version from core.

**Reply by choppsv1** [reply]
Um, ok. So I either heed this warning or skip the audit checks? I thought the system handled the namespace collision, but in this case one had to specify the full tap path and couldn't simply use the short formula name. It seems you are saying that this is not the case anymore and I should expect the system to do surprising things instead.

**Reply by gromgit** [reply]
Any formula that depends (directly or indirectly) on `emacs` will not be satisfied with `choppsv1/emacsmacport/emacs`, and will therefore force an installation of core `emacs`. That automatically results in a namespace collision.

So you can stick with `emacs-mac` and ignore the failing audit (only an issue when you need to bump its version), or rename to `emacs` and deal with Homebrew complaining every time it needs to walk a core dependency graph that passes through `emacs`. At minimum, that includes every time a dependent formulae is installed, reinstalled or upgraded.

See: https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap#naming-your-formulae-to-avoid-clashes

**Reply by choppsv1** [reply]
Thanks. Actually I already hit this problem no matter what my formula is named. It's still emacs and so it still installs in {prefix}/bin/emacs and so conflicts with the core emacs. It does satisfy the emacs requirement, but there's no way to tell homebrew that.

Anyway I'm going to just stop trying to make this CI stuff work. I was just trying to make it easier for people to use the work I did, but this is a lot of hassle to just get some unnecessary bottles built. People can still use the formula anyway.

**Reply by Gcenx** [reply]
If your only wanting to make use of the emacs.app you could make the formula keg_only.

If you need the other binary files you can add it to the front of PATH.

**Reply by choppsv1** [reply]
It's a replacement for the core emacs. I'm not sure what keg_only implies... In any case, again just trying to make it so people could easily take advantage of the notification support I added to emacs. I don't have anymore time to try and work on packaging to this end though. Thanks.

**Reply by Gcenx** [reply]
The keg_only means a formula won’t be linked into place.

Homebrew doesn’t allow overriding core formula, so if your modified version is intended to be used standalone keg_only would work.

However if the idea is to use your version of the formula as a build dependency for another core formula that won’t work, you’d need to also add those formula to your tab and modify them to use your modified formula.

**Reply by choppsv1** [reply]
I understand what your saying, but then there's people actually trying to get things done :)

https://github.com/d12frosted/homebrew-emacs-plus
https://github.com/railwaycat/homebrew-emacsmacport

The latter tap of which I forked from, both replace the core emacs recipe (i.e., link emacs into the std path), and are very popular and used by a large number of people, perhaps even more than the core formula :)

**Reply by Gcenx** [reply]
If seems your not following, installing one of these emacs formula will install and link into place as usual but the problem happens later.

After installing one of those formula (or your own) and let’s say you now also needed `proof-general` that will request the core version of emacs to be installed.

Making the formula keg_only avoids this and simply placing you own custom version at the beginning of your users PATH you can use your own version without issue. Might also need to give the .app bundle a unique name but still.

Brew doesn’t allow taps to override core formula.

**Reply by choppsv1** [reply]
I actually also maintain a variation of mu (mu-mac) for this very reason. In any case I spent a couple long days implementing a very useful addition to emacs (adding OSX native notification support) which I consider time well spent.  As the formula as written works (building from source) for me and probably others who care to use it, the hours I've spent now trying to make bottles work, doesn't seem like time well spent in comparison. So I'm going to leave things be. Thanks for the time you've spent explaining things.

**Reply by Gcenx** [reply]
True as long as it builds it doesn't make much difference, but if you did want to provide a prebuilt bottle you can skip using the GitHub action and just do it manually then add a release.

Uninstall your formula then run;
```
brew install --build-bottle emacs-mac
brew bottle emacs-mac
```
Now the generated bottle will be within `~/` you can now add a release and upload the prebuilt bottle(s) then just add the "bottle do" section as provided by "bottle"

Heres an example taken from one of my private taps.
```
bottle do
  root_url "https://github.com/Gcenx/homebrew-customllvm/releases/download/customllvm-8.0.0"
  sha256 cellar: :any, big_sur:  "a82cd740f295686f76e3357860362229b5b1ed8a3527dd89fac9ca92a440b03e"
  sha256 cellar: :any, catalina: "8f6f179400af77a99648e4030667629c985520725006fbe36506eb6621503b7b"
end
```

**Reply by choppsv1** [reply]
I did give this a shot actually, but got some worrying error messages and (sorry for sounding like a broken record) don't have time to investigate :)

Part of my code changes to emacs were to fix code-signing (required for notifications to work and perhaps also b/c I'm running on MI mac mini): https://github.com/choppsv1/emacs-mac/commit/2ee07378f56485245ddd6e3256c332bca4b93e9f For notifications to work the entire Emacs.app directory contents have to be signed, not just the Emacs binary. Homebrew is doing some code signing itself and encountering errors. So again, had to abandon this.

Example of errors:

```
(default) [08:11:49 ja:~/p]$ brew bottle emacs-mac
==> Determining choppsv1/emacsmacport/emacs-mac bottle rebuild...
==> Bottling emacs-mac--27.2-mac-1.7.arm64_big_sur.bottle.tar.gz...
Error: Failed applying an ad-hoc signature to /opt/homebrew/Cellar/emacs-mac/27.2-mac-1.7/Emacs.app/Contents/MacOS/Emacs:
/opt/homebrew/Cellar/emacs-mac/27.2-mac-1.7/Emacs.app/Contents/MacOS/Emacs: replacing existing signature
/opt/homebrew/Cellar/emacs-mac/27.2-mac-1.7/Emacs.app/Contents/MacOS/Emacs: code object is not signed at all
In subcomponent: /opt/homebrew/Cellar/emacs-mac/27.2-mac-1.7/Emacs.app/Contents/MacOS/Emacs.pdmp

Error: Failed applying an ad-hoc signature to /opt/homebrew/Cellar/emacs-mac/27.2-mac-1.7/Emacs.app/Contents/MacOS/Emacs:
/opt/homebrew/Cellar/emacs-mac/27.2-mac-1.7/Emacs.app/Contents/MacOS/Emacs: replacing existing signature
/opt/homebrew/Cellar/emacs-mac/27.2-mac-1.7/Emacs.app/Contents/MacOS/Emacs: code object is not signed at all
In subcomponent: /opt/homebrew/Cellar/emacs-mac/27.2-mac-1.7/Emacs.app/Contents/MacOS/Emacs.pdmp

... [many more]
```

**Reply by Gcenx** [reply]
Yeah that’s definite specific to the arm build as I did manage to bottle this under Rosetta2 but didn’t have chance to try an arm build before heading out to work.

If I figure it out I’ll let you know.
